PhysicsTools/NanoAOD/plugins/GenWeightsTableProducer.cc

```diff
@@ -395,6 +422,58 @@ class GenWeightsTableProducer : public edm::global::EDProducer<edm::StreamCache<
     counter->incLHE(genWeight, wScale, wPDF, wRwgt, wNamed, wPS);
   }
 
+  void fillLHEPdfWeightTablesFromGenInfo(
+					 Counter * counter,
+					 const DynamicWeightChoiceGenInfo * weightChoice,
+					 double genWeight,
+					 const GenEventInfoProduct & genProd,
+					 std::unique_ptr<nanoaod::FlatTable> & outScale,
+					 std::unique_ptr<nanoaod::FlatTable> & outPdf,
+					 std::unique_ptr<nanoaod::FlatTable> & outNamed,
+					 std::unique_ptr<nanoaod::FlatTable> & outPS ) const
+  {
+
+    const std::vector<unsigned int> & scaleWeightIDs = weightChoice->scaleWeightIDs;
+    const std::vector<unsigned int> & pdfWeightIDs   = weightChoice->pdfWeightIDs;
+    
+    auto weights = genProd.weights();
+    double w0 = weights.at(1); 
+    double originalXWGTUP = weights.at(1); ;
+    
+    std::vector<double> wScale, wPDF, wPS;
+    for (auto id : scaleWeightIDs) wScale.push_back(weights.at(id)/w0); 
+    for (auto id : pdfWeightIDs) {
+      wPDF.push_back(weights.at(id)/w0);
+    }
+    
+    unsigned int firtPSweight = 0;
+    if (weights.at(weights.size()-13)==genWeight) firtPSweight = weights.size()-13;
```

Comment:
This way of checking where the PS weights are is extremely fragile.
In cases where we only read the PS weights from the GenEventInfoProduct, we check `genProd.weights().size()` before reading the PS weights from fixed locations - still fragile, but less than this.
Would it be possible to have at least something that does not depend on the value of the weight, that might be the same also by accident in multiple places along the vector?

Reply:
ok, I now use a vector of weight ids filled from weightNames() as we do for scale and pdf   variations